ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +194,82 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestHeader.Number.Int64() <= lastStoredHeader.Number.Int64() {
+			return nil
+		}
+		numBlocksToFetch = int(latestHeader.Number.Int64() - lastStoredHeader.Number.Int64())
 	}
-	if err != nil {
-		return err
+
+	allEvents := []*Event{}
+	var syncErr error
+	for i := 0; i < numBlocksToFetch; i++ {
+		// Optimization: If numBlocksToFetch is 1, we already know what the nextHeader is, so avoid
+		// fetching it again. If there is more then 1 block to fetch, compute each from the last
+		// stored and fetch it
+		nextHeader := latestHeader
+		if numBlocksToFetch != 1 {
+			lastStoredHeader, syncErr := w.stack.Peek()
+			if syncErr != nil {
+				break
+			}
+			nextBlockNumber := big.NewInt(0).Add(lastStoredHeader.Number, big.NewInt(1))
+			nextHeader, syncErr = w.client.HeaderByNumber(nextBlockNumber)
+			if syncErr != nil {
+				break
+			}
+		}
+
+		var events []*Event
+		events, syncErr = w.buildCanonicalChain(nextHeader, events)
+		allEvents = append(allEvents, events...)
+		if syncErr != nil {
+			break
+		}
+	}
+	if len(allEvents) == 0 {
+		return syncErr
+	}
+	if w.shouldRevertChanges(lastStoredHeader, allEvents) {
+		if err := w.stack.Reset(); err != nil {
+			return err
+		}
+	} else {
+		if err := w.stack.Checkpoint(); err != nil {
+			return err
+		}
+		w.blockFeed.Send(allEvents)
 	}
-	return nil
+
+	return syncErr
+}
+
+func (w *Watcher) shouldRevertChanges(lastStoredHeader *miniheader.MiniHeader, events []*Event) bool {
+	if len(events) == 0 || lastStoredHeader == nil {
+		return false
+	}
+	// If we haven't progressed in terms of block number, revert back to previous "latest" block.
+	// This ensures block events always leave the node further ahead, preventing unnecessary thrash
+	// during block-reorgs (which tend to cluster)
+	newLatestHeader := events[len(events)-1].BlockHeader
+	return newLatestHeader.Number.Int64() <= lastStoredHeader.Number.Int64()
```

Comment:
If we just need to do a comparison and nothing else, `Cmp` is probably faster here and uses less memory.